cmd/internal/database/postgres/upgrade.go

```diff
@@ -0,0 +1,241 @@
+
+	// run the pg_upgrade command as postgres user
+	pgUser, err := user.Lookup("postgres")
+	if err != nil {
+		return err
+	}
+	uid, err := strconv.Atoi(pgUser.Uid)
+	if err != nil {
+		return err
+	}
+
+	// remove /data/postgres-new if present
+	newDataDirTemp := path.Join("/data", "postgres-new")
+	err = os.RemoveAll(newDataDirTemp)
+	if err != nil {
+		db.log.Infow("unable to remove new datadir, skipping upgrade", "error", err)
+		return nil
+	}
+
+	// initdb -D /data/postgres-new
+	cmd := exec.Command(postgresInitDBCmd, "-D", newDataDirTemp)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Credential: &syscall.Credential{Uid: uint32(uid)},
+	}
+	err = cmd.Run()
+	if err != nil {
+		db.log.Errorw("unable to run initdb on new new datadir, skipping upgrade", "error", err)
+		return nil
+	}
+	db.log.Infow("new database directory initialized")
+
+	// restore old pg_hba.conf
+	pgHBAConf, err := os.ReadFile(path.Join(db.datadir, postgresHBAConf))
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(path.Join(newDataDirTemp, postgresHBAConf), pgHBAConf, 0600)
+	if err != nil {
+		return err
+	}
+
+	newPostgresBindir, err := db.getBindir(postgresConfigCmd)
+	if err != nil {
+		return fmt.Errorf("unable to detect bindir of actual postgres %w", err)
+	}
+
+	pgUpgradeArgs := []string{
+		"--old-datadir", db.datadir,
+		"--new-datadir", newDataDirTemp,
+		"--old-bindir", oldPostgresBindir,
+		"--new-bindir", newPostgresBindir,
+		"--link",
+	}
+	db.log.Infow("running pg_upgrade with", "args", pgUpgradeArgs)
+	cmd = exec.Command(postgresUpgradeCmd, pgUpgradeArgs...) //nolint:gosec
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Credential: &syscall.Credential{Uid: uint32(uid)},
+	}
+	cmd.Dir = pgUser.HomeDir
+	err = cmd.Run()
+	if err != nil {
+		db.log.Errorw("unable to run pg_upgrade on new new datadir, abort upgrade", "error", err)
+		return fmt.Errorf("unable to run pg_upgrade %w", err)
+	}
+	db.log.Infow("pg_upgrade done")
+
+	// rm -rf /data/postgres
+	err = os.RemoveAll(db.datadir)
+	if err != nil {
+		return fmt.Errorf("unable to remove old datadir %w", err)
+	}
+
+	err = os.Rename(newDataDirTemp, db.datadir)
+	if err != nil {
+		return fmt.Errorf("unable to rename upgraded datadir to destination, a full restore is required, error %w", err)
+	}
+
+	db.log.Infow("pg_upgrade done and new data in place", "took", time.Since(start))
+
+	return nil
+}
+
+// Helpers
+
+func (db *Postgres) getBinaryVersion(pgConfigCmd string) (int, error) {
+	// pg_config  --version
+	// PostgreSQL 12.16
+	cmd := exec.Command(pgConfigCmd, "--version")
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return 0, fmt.Errorf("unable to detect postgres binary version, skipping upgrade %w", err)
+	}
+	_, binaryVersionString, found := strings.Cut(string(out), "PostgreSQL ")
```

Comment:
Suggestion: https://go.dev/play/p/5Dmo2z4Ki-K

Reply:
thanks for the tip !